File: two-sum-iii-data-structure-design/two_sum170.go
// https://leetcode.com/problems/two-sum-iii-data-structure-design/

package main

import (
	"fmt"
	"sort"
)

type TwoSum struct {
	nums   []int
	sorted bool
}

func Constructor() TwoSum {
	return TwoSum{
		nums:   make([]int, 0),
		sorted: false,
	}
}

// T: O(1), S: O(n)
func (ts *TwoSum) Add(number int) {
	ts.nums = append(ts.nums, number)
	ts.sorted = false
}

// T: O(n log n), S: O(n)
func (ts *TwoSum) Find(value int) bool {
	if !ts.sorted {
		sort.Ints(ts.nums)
		ts.sorted = true
	}
	left, right := 0, len(ts.nums)-1
	for left < right {
		n1, n2 := ts.nums[left], ts.nums[right]
		sum := n1 + n2
		if sum < value {
			left++
		} else if sum > value {
			right--
		} else {
			return true
		}
	}
	return false
}

func main() {
	obj := Constructor()
	obj.Add(1)
	obj.Add(3)
	obj.Add(5)
	fmt.Println(obj.Find(4)) // true
	fmt.Println(obj.Find(7)) // false
}

File: two-sum-iii-data-structure-design/TwoSum170.java
// https://leetcode.com/problems/two-sum-iii-data-structure-design/

import java.util.*;

class TwoSum170 {

    private List<Integer> nums;
    private boolean sorted;

    public TwoSum170() {
        nums = new ArrayList<>();
        sorted = false;
    }

    // T: O(1), S: O(n)
    public void add(int number) {
        nums.add(number);
        sorted = false;
    }

    // T: O(n log n), S: O(n)
    public boolean find(int value) {
        if (!sorted) {
            Collections.sort(nums);
            sorted = true;
        }
        int left = 0, right = nums.size() - 1;
        while (left < right) {
            int n1 = nums.get(left);
            int n2 = nums.get(right);
            int sum = n1 + n2;
            if (sum < value) {
                left++;
            } else if (sum > value) {
                right--;
            } else {
                return true;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        var ts = new TwoSum170();
        ts.add(1);
        ts.add(3);
        ts.add(5);
        System.out.println(ts.find(4)); // true
        System.out.println(ts.find(7)); // false
    }

}
